### PR TITLE
Support for non-styling custom attribute

### PR DESCRIPTION
**Pull request by ritheshSalyan**
This pull request contains the following major changes:

1. Change type of SizeAttribute from String to double.
2. support for the non-styling custom attribute.

### CODE REVIEW

File: app/assets/sample_data.json [hunk 1]
```diff
@@ -451,7 +451,7 @@
    },
    {
       "attributes": {
-         "size": "small"
```
Comment (singerdmx) [inline app/assets/sample_data.json:454]: Can you keep those but add three inserts?

**Review by singerdmx — COMMENTED** [review]
I thought you are supposed to use stable branch.
That is stable channel

File: app/windows/.gitignore [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+# Visual Studio build-related files.
+x64/
+x86/
+
+# Visual Studio cache files
+# files ending in .cache can be ignored
+*.[Cc]ache
+# but keep track of directories ending in .cache
+!*.[Cc]ache/
```
Comment (singerdmx) [inline app/windows/.gitignore:17]: Should .vscode/launch.json  be ignored?

Comment (ritheshSalyan) [reply app/windows/.gitignore:17]: you can ignore that

File: example/main.dart [hunk 3]
```diff
@@ -1,40 +0,0 @@
-import 'dart:async';
-import 'dart:io';
-
```
Comment (singerdmx) [inline example/main.dart:3]: This file is for pub. Why are you removing it?

File: lib/models/documents/attribute.dart [hunk 4]
```diff
@@ -201,12 +202,12 @@ class StrikeThroughAttribute extends Attribute<bool> {
   StrikeThroughAttribute() : super('strike', AttributeScope.INLINE, true);
 }
 
-class FontAttribute extends Attribute<String> {
-  FontAttribute(String val) : super('font', AttributeScope.INLINE, val);
+class FontAttribute extends Attribute<double> {
```
Comment (singerdmx) [inline lib/models/documents/attribute.dart:205]: We want to support both string and double, right?

File: lib/models/rules/format.dart [hunk 5]
```diff
@@ -132,3 +132,41 @@ class ResolveInlineFormatRule extends FormatRule {
...
   }
 }
+
+class ResolveInlineIgnoreFormatRule extends FormatRule {
+  const ResolveInlineIgnoreFormatRule();
+
+  @override
+  Delta applyRule(Delta document, int index,
+      {int len, Object data, Attribute attribute}) {
+    if (attribute.scope != AttributeScope.IGNORE) {
```
Comment (singerdmx) [inline lib/models/rules/format.dart:142]: Should this be `==` ?

File: lib/models/rules/rule.dart [hunk 6]
```diff
@@ -70,6 +71,6 @@ class Rules {
         throw e;
       }
     }
-    throw ('Apply rules failed');
+    // throw ('Apply rules failed');
```
Comment (singerdmx) [inline lib/models/rules/rule.dart:74]: Can we do something smarter?
Apply a rule that does nothing for ignore

File: lib/models/documents/attribute.dart (hunk 4 above)
Comment (ritheshSalyan) [reply lib/models/documents/attribute.dart:205]: yes.

**Review by singerdmx — COMMENTED** [review]
If your goal is support font size of double, you can still assume string at the beginning.
Parse the string to see if it is double and do logic accordingly.
There seems no need to play with ignore or rule

File: lib/models/documents/attribute.dart (hunk 4 above)
Comment (singerdmx) [reply lib/models/documents/attribute.dart:205]: Then don't change it to double.
We still want to accept string like `small`

File: lib/models/rules/rule.dart (hunk 6 above)
Comment (ritheshSalyan) [reply lib/models/rules/rule.dart:74]: yes, you can. I just copied the same rule as of inline  `ResolveInlineFormatRule()`

File: lib/models/documents/attribute.dart (hunk 4 above)
Comment (singerdmx) [reply lib/models/documents/attribute.dart:205]: > If your goal is support font size of double, you can still assume string at the beginning.
> Parse the string to see if it is double and do logic accordingly.
> There seems no need to play with ignore or rule

image


You just need to handle when `font.value` is `double`, right?

File: app/assets/sample_data.json (hunk 1 above)
Comment (ritheshSalyan) [reply app/assets/sample_data.json]: sure. 
check the latest commits.
I reverted type of SizeAttribute to string

now you can set font size in 2 ways 

```dart
   double fontSize = 50.0;
 _controller.formatSelection(SizeAttribute("$fontSize"));

```

Or
```dart

 _controller.formatSelection(SizeAttribute("small"));

```

**Review by singerdmx — COMMENTED** [review]
Can we have a much smaller change that is only handling font value of double, please?
And probably against stable?
This change is too big

File: lib/widgets/text_line.dart [hunk 7]
```diff
@@ -148,6 +149,11 @@ class TextLine extends StatelessWidget {
           res = res.merge(defaultStyles.sizeHuge);
           break;
         default:
+          double fontSize = double.tryParse(size.value);
```
Comment (singerdmx) [inline lib/widgets/text_line.dart:152]: I think you only need this file's change, right?

File: app/assets/sample_data.json [hunk 8]
```diff
@@ -473,6 +473,30 @@
       },
       "insert": "Huge"
    },
+   {
```
Comment (singerdmx) [inline app/assets/sample_data.json:476]: Yes, we need this as well

File: lib/widgets/toolbar.dart [hunk 9]
```diff
@@ -742,6 +742,7 @@ class _HistoryButtonState extends State<HistoryButton> {
   }
 
   void _setIconColor() {
+    if(!mounted) return;
```
Comment (singerdmx) [inline lib/widgets/toolbar.dart:745]: OK, we can keep this

**Review by singerdmx — COMMENTED** [review]
can you shrink down to three files only?
Make a separate pull request if you want change others

**Comment by ritheshSalyan** [comment]
ok

**Comment by singerdmx** [comment]
@ritheshSalyan  are you planning to send a new MR?

**Pull request closed without merging**